File: 0135-candy/0135-candy.java
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int[] candies = new int[n];
        Arrays.fill(candies, 1); // start by giving each child one candy

        // iterate over the ratings array from left to right
        for (int i = 1; i < n; i++) {
            if (ratings[i] > ratings[i - 1]) {
                candies[i] = candies[i - 1] + 1; // give the current child one more candy than its left neighbor
            }
        }

        // iterate over the ratings array from right to left
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                candies[i] = Math.max(candies[i], candies[i + 1] + 1); // give the current child one more candy than its right neighbor
            }
        }

        // sum up the candies for each child
        int totalCandies = 0;
        for (int i = 0; i < n; i++) {
            totalCandies += candies[i];
        }

        return totalCandies;
    }
}

File: 0135-candy/0135-candy.go
func candy(ratings []int) int {
    length := len(ratings)
    candies := make([]int, length)
    for i := range candies {
        candies[i] = 1
    }
    
    for i := 1; i < length; i++ {
        if ratings[i] > ratings[i - 1] {
            candies[i] = candies[i - 1] + 1
        }
    }
    
    for i := length - 2; i > -1; i-- {
        if ratings[i] > ratings[i + 1] {
            candies[i] = max(candies[i], candies[i + 1] + 1)
        }
    }
    
    result := 0
    for _, v := range candies {
        result += v
    }
    
    return result
}

func max(a int, b int) int {
	if a > b {
		return a
	}

	return b
}